client/lib/validation.rs

```diff
@@ -164,14 +164,23 @@ pub(crate) fn validate_query_response(
         }
     }
 
-    core::validate_query_proof(
-        &state_root_hash.to_owned().into(),
-        &proofs,
-        key,
-        path,
-        proof_value,
-    )
-    .map_err(Into::into)
+    let state_root_hash: Digest = {
+        let value = object
+            .get(GET_ITEM_RESULT_STATE_ROOT_HASH)
+            .ok_or(ValidateResponseError::ValidateResponseFailedToParse)?;
+
+        let value_str = value
+            .as_str()
+            .ok_or(ValidateResponseError::ValidateResponseFailedToParse)?;
+
+        let value_bytes = hex::decode(value_str)
+            .map_err(|_| ValidateResponseError::ValidateResponseFailedToParse)?;
+
+        bytesrepr::deserialize(value_bytes)?
+    };
+
+    core::validate_query_proof(&state_root_hash.into(), &proofs, key, path, proof_value)
```

Comment:
This is sadly more complicated now.

You have to do a few things to check integrity now:

    1. Get the block header and check it has the state root provided.
    2. Hash the block header and check it matches what was queried.
    3. Validate the proof like this

Sadly, it kind of breaks security to skip the first two steps which is a pain :(

We have an implementation of [CEP 41][1] on a branch.  Once that lands we could replace steps 1 and 2 with another kind of Merkle proof.

[1]: https://github.com/casper-network/ceps/blob/master/text/0041-extensible-block-body.md